app/questionnaire/path_finder.py

```diff
@@ -204,15 +207,42 @@ def _get_next_block_id(self, rule):
             return self.schema.get_first_block_id_for_group(rule["group"])
         return rule["block"]
 
-    def _remove_rule_answers(self, goto_rule, this_location):
-        # We're jumping backwards, so need to delete all answers from which
-        # route is derived. Need to filter out conditions that don't use answers
-        if "when" in goto_rule.keys():
-            for condition in goto_rule["when"]:
-                if "meta" not in condition.keys():
-                    self.answer_store.remove_answer(condition["id"])
+    def _remove_current_blocks_answers_for_backwards_routing(
+        self, rules, this_location
+    ):
+        answer_ids_for_current_block = self.schema.get_answer_ids_for_block(
+            this_location.block_id
+        )
+        if "when" in rules.keys():
```

Comment:
Don't need to use `keys`, it is faster without using it 
```suggestion
        if "when" in rules:
```